### ISSUE TITLE
limit_train_batches or limit_train_samples ?

### ISSUE THREAD

**Issue by Hannibal046** [Lightning-AI/lightning#15595]
### Bug description

Hi, I notice that the `limit_train_batches` is actually `limit_train_samples`. Because the `real_batch_size  = batch_size * num_gpus * gradient_accumulation`, so maybe `limit_train_samples` would be less vague ?

**Comment by awaelchli** [comment]
@Hannibal046 No, it is not like you describe. `limit_train_batches` is really what it says, it limits the number of batches that are fetched from the dataloader! 

Please note that gradient accumulation is NOT exactly the same as training with a larger batch size. 

I'm not saying that `limit_train_samples` is not useful, but the existing `limit_train_batches` is IMO pretty well defined in Lightning. 

Since we are most of the time using PyTorch's DataLoader which does batching, we can't really control the number of samples so easily. We can't set the DataLoader batch size to a fixed number and then stop after EXACTLY `limit_train_samples`. I think it would be more confusing to users, especially because there is also an option `drop_last` in the DataLoader. 

Hope this helps clear things up for you! Let me know if you have any further questions.

**Comment by Hannibal046** [comment]
Thanks so much ! I got it. But I don't quite understand this: 
> Please note that gradient accumulation is NOT exactly the same as training with a larger batch size.

BTW, I have another two questions:
1. if I give `max_steps` to `Trainer`, can I get corresponding `max_epochs` as trainer attributes ?
2. In the manual optimization example with gradient accumulation, should it be like this: `self.manual_backward(loss/N)` ? 
```
def __init__(self):
    super().__init__()
    self.automatic_optimization = False


def training_step(self, batch, batch_idx):
    opt = self.optimizers()

    loss = self.compute_loss(batch)
    self.manual_backward(loss)

    # accumulate gradients of N batches
    if (batch_idx + 1) % N == 0:
        opt.step()
        opt.zero_grad()
```

**Comment by awaelchli** [comment]
> Thanks so much ! I got it. But I don't quite understand this:
> 
> Please note that gradient accumulation is NOT exactly the same as training with a larger batch size.


Think about what happens to the weights of the network during gradient accumulation and then compare that to the regular training:

**Accumulation=3:**
batch0 -> compute gradient -> batch1 -> compute gradient -> batch2 -> compute gradient -> sum gradients -> updated weights

vs.

**3 regular training steps:**
batch0 -> compute gradient -> updated weights-> batch1 -> compute gradient -> update weights -> batch2 -> compute gradient -> update weights 

We can observe that the optimization is different, leading to different weights at the end even though the seen batches were the same!


> if I give max_steps to Trainer, can I get corresponding max_epochs as trainer attributes

No. max_steps and max_epochs are unrelated. You can set both if you want.


> In the manual optimization example with gradient accumulation, should it be like this: self.manual_backward(loss/N) ?

No.

**Comment by Hannibal046** [comment]
Hi, thanks for reply. In my opinion, I think gradient accumulation is actually equivalent to large batch size, and here is an example. The net_1 and net_2 actually get same gradient ( with loss/N):
![draft ipynb — selfmem  SSH: 2560  2022-11-13 12-09-24](https://user-images.githubusercontent.com/38466901/201505273-d9fdb66a-3b10-4021-a334-8d31d69295d1.png)
And here is the code:
```
from copy import deepcopy
import torch
x = [torch.tensor([i],dtype=torch.float32) for i in range(3)]
y = [torch.tensor(i+1) for i in range(3)]
net_1 = torch.nn.Linear(1,1)
net_2 = deepcopy(net_1)
print(net_1.weight.data,net_2.weight.data)
print(net_1.weight.data.data_ptr()==net_2.weight.data.data_ptr())

## batch_size = 3, gradient_accumulation_steps = 1
loss  = torch.mean(torch.stack(y) - net_1(torch.stack(x)))
print(loss)
loss.backward()
print("grad:",net_1.weight.grad)

## batch_size = 1, gradient_accumulation_steps = 3
for idx in range(3):
    loss = y[idx] - net_2(x[idx])
    print(loss)
    loss = loss / 3
    loss.backward()
print("grad:",net_2.weight.grad)
```

**Comment by awaelchli** [comment]
>  In my opinion, I think gradient accumulation is actually equivalent to large batch size

How gradient accumulation works is not an opinion. It is very well defined. I encourage you to look up the definition and at code samples of how it is implemented in practice (just Google it and you will find the resources :)). 

Your code sample doesn't show anything interesting. You never do anything with your gradients. The weights never get  updated. This doesn't show anything. 

I will let you think about this a bit more. If after that you still struggle, I can provide an example.

**Comment by Hannibal046** [comment]
Same gradient will give same optimization results:
```python
from copy import deepcopy
import torch
x = [torch.tensor([i],dtype=torch.float32) for i in range(3)]
y = [torch.tensor(i+1) for i in range(3)]
net_1 = torch.nn.Linear(1,1)
optim_1 = torch.optim.SGD(net_1.parameters(),lr=0.01)
net_2 = deepcopy(net_1)
optim_2 = torch.optim.SGD(net_2.parameters(),lr=0.01)
print("before optimization:")
print("net_1 weight:",net_1.weight.data)
print("net_2 weight:",net_2.weight.data)

## batch_size = 3
loss  = torch.mean(torch.stack(y) - net_1(torch.stack(x)))
loss.backward()
optim_1.step()
print("after optimization:")
print("net_1 grad:",net_1.weight.grad,'\n',"net_1 weight:",net_1.weight.data)

## batch_size = 1, gradient_accumulation_steps = 3
for idx in range(3):
    loss = y[idx] - net_2(x[idx])
    loss = loss / 3
    loss.backward()
optim_2.step()
print("net_2 grad:",net_2.weight.grad,'\n',"net_2 weight:",net_2.weight.data)
```
This is the output:
```
before optimization:
net_1 weight: tensor([[0.3543]])
net_2 weight: tensor([[0.3543]])
after optimization:
net_1 grad: tensor([[-1.]]) 
 net_1 weight: tensor([[0.3643]])
net_2 grad: tensor([[-1.]]) 
 net_2 weight: tensor([[0.3643]])
```

And this is the official code by Pytorch [link](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation)
```
scaler = GradScaler()

for epoch in epochs:
    for i, (input, target) in enumerate(data):
        with autocast(device_type='cuda', dtype=torch.float16):
            output = model(input)
            loss = loss_fn(output, target)
            loss = loss / iters_to_accumulate

        # Accumulates scaled gradients.
        scaler.scale(loss).backward()

        if (i + 1) % iters_to_accumulate == 0:
            # may unscale_ here if desired (e.g., to allow clipping unscaled gradients)

            scaler.step(optimizer)
            scaler.update()
            optimizer.zero_grad()
```

**Comment by awaelchli** [comment]
I'm sorry, maybe there is a misunderstanding somewhere. Can you explain again why limit_train_samples is needed and why

> so maybe limit_train_samples would be less vague

How would you implement this? Please show a training loop as a concrete example of the stopping condition.

**Comment by Hannibal046** [comment]
Hi, I think this depends on how to define **batch**. I am more familiar with this definition:
```
The
```

**Comment by riklopfer** [comment]
@awaelchli I just wanted to confirm that when `limit_train_batches` is set, that allocation of batches is distributed between GPUs such that each device would see  `limit_train_batches // num_gpus` batches.
